### ISSUE TITLE
improved devcontainer support?

### ISSUE THREAD

**Issue by cboettig** [rocker-org/rocker-versioned2#520]
Looking at the amazing work @eitsupi and @revodavid have done with devcontainer standard (particularly for deploy on codespaces, which looks rather promising in some contexts), I'm wondering if we should support more of that boilerplate here.  Not sure if that would b best as an additional image in the stack or (probably better) an additional bash script in `rocker_scripts` that can bundle up at least some of the additional things that get added to the [devcontainer Dockerfile](https://github.com/revodavid/devcontainers-rstudio/blob/main/.devcontainer/Dockerfile) and [debian script](https://github.com/revodavid/devcontainers-rstudio/blob/main/.devcontainer/library-scripts/common-debian.sh).  

(There's also a lot in there I haven't quite wrapped my head around or am unsure about.  e.g. see https://github.com/revodavid/devcontainers-rstudio/issues/1, I'm not sure that we want to use separate users `vscode` and `rstudio` on the same VM? seems to create a minor issue for git too?)  Thoughts / ideas / objections welcome!

**Comment by eitsupi** [comment]
First of all, I think it is necessary for you to understand the status of devcontainer (microsoft/vscode-dev-containers#1589).

For a long time, devcontainer could only be handled from the proprietary vscode extension available only on Visual Studio Code, but Microsoft has recently moved this to an open source implementation.
https://github.com/devcontainers/spec
https://github.com/devcontainers/cli

In parallel, they are working on implementing a feature called devcontainer features, which will add functionality on the containers.
https://github.com/devcontainers/features

This is effectively a more generalized version of the Rocker Scripts that exist in this repository.
In other words, I believe that almost all of the contents of this repository could be migrated to the devcontainer features.
(devcontainer features spec still a work in progress and subject to destructive changes / I was going to try this in the near future in my personal repository)
If such a thing could be done, it would more generally accomplish what I was trying to do in #483.

For example, the contents of [the script](https://github.com/devcontainers/features/blob/f82307ad60514d1e912a9aa2b64204721fe955eb/src/common-utils/install.sh) can be imported into any debian-based container by building devcontainer from the following json file.

```json
"name": "my-project-devcontainer",
"image": "mcr.microsoft.com/devcontainers/base:ubuntu",  // Any generic, debian-based image.
"features": {
    "ghcr.io/devcontainers/features/common-utils:1": {
        "version": "latest"
    }
}
```

As devcontainer featuers spread, shell scripts for installing any software should be able to be shared among arbitrary Debian base images, eliminating the need to add shell scripts in the repository for building containers on a dare.

Also, in relation to devcontainer's move to OSS, it would make sense to move [the R definition](https://github.com/microsoft/vscode-dev-containers/tree/main/containers/r) to the Rocker project, since https://github.com/microsoft/vscode-dev-containers is scheduled to be closed in the near future.

> I'm not sure that we want to use separate users `vscode` and `rstudio` on the same VM?

Yes, that is not necessary, see revodavid/devcontainers-rstudio#2.
(@revodavid contacted me personally a few months ago and I have had time to look through that repository before)

**Comment by cboettig** [comment]
Thanks @eitsupi , this is super as always.  Definitely helps to understand a bit more of the big picture goals Microsoft has with devcontainers, and glad to see they have open-sourced the project.  (Though is the devcontainer.json yet consumed by any third-party tooling (i.e. outside of vscode / codespaces?)  I think I understand and like their distinction between 'development' and 'production' containers in the [overview page](https://containers.dev/overview), but still not really sure I see quite how devcontainers fits within the larger ecosystem of building & orchestrating container-based workflows.  Is a `devcontainer.json` file operating like our `*.json` files in `stack/`?  How does it compare to existing orchestration tools, like kubernetes YAML files?  (to me, I see the need to write a devcontainer.json to deploy on Microsoft systems, i.e. VSCode and now codespaces, but imagine that one would accomplish something quite similar on a different provider's platform with some different orchestration metadata, like some k8s yaml.... From your comments I'm encouraged that devcontainer.json might be something both more platform agnostic and solving some additional needs beyond deploy, but I haven't wrapped my head around precisely what that is....)


More immediately, I'd be happy to see the devcontainer R-definition move here if the existing container is to be archived.  Thanks for linking that PR too, agree :100: on that.  I like @revodavid's example repo, but it feels like that repo has a size-able code base in it that could be moved upstream (e.g. here / part of the "R definition"), rather than be stuff every single repo needs to duplicate, right?

**Comment by cboettig** [comment]
Thanks, this is all helpful.  

Right, I've noticed the delay in deploy, and also that each of these repos is building a 50+ line Dockerfile that, among other things, involves running a ~ 450 line debian-common.sh bash script.  It seems that it would be desirable to have much of that install / config prebuilt into an upstream rocker container, both to achieve faster deploys and also to avoid duplicating code that could then diverge.  Individual repos would I guess still contain a Dockerfile as well as the devcontainer.json file, but it could be one that was considerably more concise then, right?  

I guess the next step would be to bring the devcontainer R-definition in over here?

**Comment by eitsupi** [comment]
@cboettig  We can now manage container configuration with json files, as in this repository, and GitHub Actions can build and push.
(Reporting capabilities and automatic updates can be added later if needed, but I believe it is not a priority for now.)
https://github.com/eitsupi/r-devcontainer-images/tree/a2c35937cc063bf107ffb557bcf75e81f15990cc

Future, `bspm` based containers can be easily added.

I am wondering if this can be brought to rocker-org.
Of course, the devcontainer cli is not yet stable, so this repo and containers should be considered experimental (I can add a badge later)

**Comment by cboettig** [comment]
Thanks @eitsupi , this is great.  All right with you to transfer the repo to rocker-org and see what we can do with it?  

Would it be possible to change the configuration such that the RStudio instance opens in the working directory of the cloned repository as an 'open project', thus allowing RStudio's git panel to also accurately reflect the github repo from which the devcontainer was launched?  Currently I think that issue is the biggest ergonomic barrier to making the devcontainer practical for RStudio-based training events etc.

**Comment by eitsupi** [comment]
> All right with you to transfer the repo to rocker-org and see what we can do with it?

👍


> Would it be possible to change the configuration such that the RStudio instance opens in the working directory of the cloned repository as an 'open project'

It appears that the default working directory can be changed by editing `/home/rstudio/.config/rstudio/rstudio-prefs.json`.

```json
{
    "initial_working_directory": "/workspace"
}
```

If this folder does not exist, it will fall back to `/home/rstudio`, so there seems to be no harm in this setting even if the path to bind mount is changed.

However, the current rocker/rstudio uses an old configuration file, so `/home/rstudio/.config/rstudio/rstudio-prefs.json` is automatically generated when RStudio is first started and cannot be edited before RStudio is started.

https://github.com/rocker-org/rocker-versioned2/blob/9423d182aefc8ddc164f030ce583ec6b6fbddb66/scripts/default_user.sh#L16-L20

Fortunately, RStudio started using the new configuration file in the `rocker/rstudio:4.0.0` era (rstudio/rstudio#5069), so we can edit the above line to configure it to use `/home/rstudio/.config/rstudio/rstudio-prefs.json` instead of `/home/rstudio/.rstudio/monitored/user-settings`.
I will send a PullRequest later.

**Comment by eitsupi** [comment]
Hi @thohan88, thanks to sharing that!

> Would love to hear if you have any experience with other approaches, @eitsupi

I was taking the approach of using docker compose as I wrote in this Japanese article.
https://qiita.com/eitsupi/items/ae0f89266b560b4e7096

The config files look like these:

```json
{
  "name": "rocker_rstudio",
  "dockerComposeFile": [],
  "service": "rstudio",
  "workspaceFolder": "/home/rstudio/works",
  "settings": {
    "r.rterm.linux": "/usr/local/bin/radian",
    "r.bracketedPaste": true,
    "r.plot.useHttpgd": true,
    "[r]": {
      "editor.wordSeparators": "`~!@#%$^&*()-=+[{]}\\|;:'\",<>/?"
    }
  },
  "extensions": ["REditorSupport.r"],
  "shutdownAction": "none",
  "remoteUser": "rstudio"
}
```

```yml
services:
  rstudio:
    build:
      context: ./.devcontainer
    container_name: rocker_rstudio
    ports:
      - "8787:8787"
    volumes:
      - ./.rstudio_config:/home/rstudio/.config
      - ./works:/home/rstudio/works
```

Instead of changing RStudio's settings, you can set `workspaceFolder` to match the directory where VSCode opens to RStudio.

The downside of this approach is more configuration files and RStudio is always running and consuming resources...

**Comment by eitsupi** [comment]
Now experimental docker image `ghcr.io/rocker-org/devcontainer/tidyverse:4` is available.
https://github.com/rocker-org/devcontainer-images/pkgs/container/devcontainer%2Ftidyverse

This image has `radian`, `languageserver`, and `httpgd` already installed, which are necessary for using `vscode-R`, and the common utilities are also installed, so we can use it comfortably without local builds.

```json
{
  "image": "ghcr.io/rocker-org/devcontainer/tidyverse:4",
  "customizations": {
    "vscode": {
      "settings": {
        "r.rterm.linux": "/usr/local/bin/radian",
        "r.bracketedPaste": true,
        "r.plot.useHttpgd": true,
        "[r]": {
          "editor.wordSeparators": "`~!@#%$^&*()-=+[{]}\\|;:'\",<>/?"
        }
      },
      "extensions": ["reditorsupport.r", "rdebugger.r-debugger"]
    }
  },
  "remoteUser": "rstudio"
}
```

@cboettig Do you think you can close this in favor of https://github.com/rocker-org/devcontainer-images?

**Comment by cboettig** [comment]
definitely!

closing in favor of further development under https://github.com/rocker-org/devcontainer-images
